File: language-testing/tcp_echo_client/tcp_echo_client.rb
#! /usr/bin/env ruby

require 'socket'

def send_receive_data hostname, port, content
  TCPSocket.open hostname, port do |socket|
    socket.puts content
    puts socket.gets
  end
end

def die msg
  $stderr.puts "Usage: #{$0} hostname port\nError: #{msg}" 
  exit 1
end

if __FILE__ == $0
  case ARGV.length
  when 0 then die "No hostname given!"
  when 1 then die "No port given!"
  when 2 then
    die "Port contains non-numbers!" unless /^\d+$/ =~ ARGV[1]
    send_receive_data ARGV[0], ARGV[1].to_i, $stdin.read
  else die "Too many arguments given!"
  end
end

File: language-testing/tcp_echo_client/tcp_echo_client.pl
#! /usr/bin/env perl

use strict;
use warnings;
use IO::Socket;

sub send_and_receive_data {
  my $sock = new IO::Socket::INET (
    PeerAddr => $_[0],
    PeerPort => $_[1],
    Proto => "tcp"
    );
  die "Failed to connect: $!\n" unless $sock;
  print $sock $_[2];
  while (<$sock>) {
    print $_;
  }
  close($sock);
}

my $usage = "Usage: $0 hostname port";
die "$usage\nError: No hostname given!\n" unless $ARGV[0];
die "$usage\nError: No port given!\n" unless $ARGV[1];
die "$usage\nError: Port contains non-digits!\n" unless $ARGV[1] =~ /^\d+$/;
&send_and_receive_data($ARGV[0], $ARGV[1], <STDIN>);
